File: app/src/main/java/com/arcinteractive/spaces/data/model/SpacePoll.kt
package com.arcinteractive.spaces.data.model

import java.util.Date

data class SpacePollOption(
    val id: String,
    val text: String
)

data class SpacePollVote(
    val id: String,
    val userId: String,
    val optionIds: List<String>,
    val createdAt: Date?,
    val updatedAt: Date?
)

data class SpacePoll(
    val id: String,
    val spaceId: String,
    val question: String,
    val options: List<SpacePollOption>,
    val createdBy: String,
    val createdByName: String,
    val createdAt: Date?,
    val updatedAt: Date?,
    val closesAt: Date?,
    val allowMultipleVotes: Boolean,
    val anonymous: Boolean,
    val deleted: Boolean,
    val votes: List<SpacePollVote> = emptyList()
) {
    val isClosed: Boolean
        get() = closesAt?.let { !it.after(Date()) } ?: false

    val totalVotes: Int
        get() = votes.size

    fun votesCount(optionId: String): Int = votes.count { it.optionIds.contains(optionId) }

    fun percentage(optionId: String): Float {
        if (totalVotes == 0) return 0f
        return votesCount(optionId).toFloat() / totalVotes.toFloat()
    }

    fun selectedOptionIds(userId: String?): Set<String> {
        if (userId.isNullOrBlank()) return emptySet()
        return votes.firstOrNull { it.userId == userId }?.optionIds?.toSet() ?: emptySet()
    }
}

File: iOS/Spaces V1.6/Models/SpacePoll.swift
import Foundation

struct SpacePollOption: Identifiable, Hashable {
    let id: String
    let text: String
}

struct SpacePollVote: Identifiable, Hashable {
    let id: String
    let userID: String
    let displayName: String?
    let emojiAvatar: String?
    let optionIDs: [String]
    let createdAt: Date?
    let updatedAt: Date?
}

struct SpacePoll: Identifiable, Hashable {
    let id: String
    let spaceID: String
    let question: String
    let options: [SpacePollOption]
    let createdBy: String
    let createdByName: String
    let createdAt: Date?
    let updatedAt: Date?
    let closesAt: Date?
    let allowMultipleVotes: Bool
    let anonymous: Bool
    let deleted: Bool
    var votes: [SpacePollVote] = []

    var isClosed: Bool {
        guard let closesAt else { return false }
        return closesAt <= Date()
    }

    var totalVotes: Int {
        votes.count
    }

    func votesCount(for optionID: String) -> Int {
        votes.filter { $0.optionIDs.contains(optionID) }.count
    }

    func percentage(for optionID: String) -> Double {
        guard totalVotes > 0 else { return 0 }
        return Double(votesCount(for: optionID)) / Double(totalVotes)
    }

    func selectedOptionIDs(for userID: String?) -> Set<String> {
        guard let userID else { return [] }
        return Set(votes.first(where: { $0.userID == userID })?.optionIDs ?? [])
    }
}
